week9/answers.sql

```diff
@@ -1,3 +1,172 @@
 -- Create your tables, views, functions and procedures here!
 CREATE SCHEMA social;
 USE social;
+
+CREATE TABLE users (
+	user_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    first_name VARCHAR(30) NOT NULL,
+    last_name VARCHAR(30) NOT NULL,
+    email VARCHAR(45) NOT NULL,
+    created_on DATETIME
+);
+
+CREATE TABLE sessions (
+	session_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED,
+    created_on TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_on TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(),
+    CONSTRAINT sessions_fk_users
+		FOREIGN KEY (user_id)
+        REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+CREATE TABLE friends (
+	user_friend_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED,
+    friend_id INT UNSIGNED,
+    CONSTRAINT friends_fk_users
+		FOREIGN KEY (user_id)
+        REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+	CONSTRAINT friends_fk_users2
+		FOREIGN KEY (friend_id)
+        REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+CREATE TABLE posts (
+	post_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED,
+    created_on TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_on TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(),
+    content VARCHAR(100),
+    CONSTRAINT posts_fk_users
+		FOREIGN KEY (user_id)
+        REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+CREATE TABLE notifications(
+	notification_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED,
+    post_id INT UNSIGNED,
+    CONSTRAINT notifications_fk_users
+		FOREIGN KEY (user_id)
+        REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+	CONSTRAINT notifications_fk_posts
+		FOREIGN KEY (post_id)
+		REFERENCES posts (post_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+CREATE OR REPLACE VIEW notification_posts AS
+    SELECT n.user_id, u.first_name, u.last_name, p.post_id, p.content
+        FROM posts p
+        INNER JOIN notifications n
+		ON p.post_id = n.post_id
+	LEFT OUTER JOIN users u
+		ON p.user_id = u.user_id;
+
+DELIMITER ;; 
+
+CREATE TRIGGER new_user
+    AFTER INSERT ON users
+    FOR EACH ROW
+BEGIN
+    DECLARE not_new_user INT UNSIGNED;
+    DECLARE recent_post INT UNSIGNED;
+    DECLARE row_not_found TINYINT DEFAULT FALSE;
+    
+    DECLARE user_cursor CURSOR FOR
+	SELECT u.user_id
+	    FROM users u
+	WHERE u.user_id != NEW.user_id;
+            
+    DECLARE CONTINUE HANDLER FOR NOT FOUND
+	SET row_not_found = TRUE;
+    
+    -- Creates the user joined posts
+    SET @new_content = CONCAT(NEW.first_name, " ", NEW.last_name, " just joined!");
```

Comment:
You won't want to use a user-defined variable, here. These are intended for global use with connections and live for the life of the connection. I'm not certain, but I think if you had other procedures/triggers/events/functions using `@new_content` that you could get a race condition (I would need to do some careful experiments to confirm, as this is undocumented behavior).

You can either DECLARE a local variable and assign the value, or simply make it a part of your INSERT statement, below.